### PR TITLE
Suspend thread pool

### PR DESCRIPTION
**Pull request by msimberg**
This adds a `suspend` method to `thread_pool_base`/`scheduled_thread_pool`. It waits for all work to be done on the pool before suspending.

### CODE REVIEW

File: hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp [hunk 1]
```diff
@@ -316,13 +316,40 @@ namespace hpx { namespace threads { namespace detail
...
+                get_background_thread_count())
+            {
+                hpx::this_thread::suspend();
+            }
+        }
+        else
+        {
+            while (sched_->Scheduler::get_thread_count() >
+                get_background_thread_count())
+            {}
```
Comment (hkaiser) [inline hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: Should this at least call `sleep(0)` instead of unconditionally spinning? Alternatively, we could re-spawn the resume function on an HPX thread instead of spinning, not sure...

File: hpx/runtime/threads/detail/thread_pool_base.hpp [hunk 2]
```diff
@@ -80,7 +80,8 @@ namespace hpx { namespace threads { namespace detail
         virtual void stop(
             std::unique_lock<compat::mutex>& l, bool blocking = true) = 0;
 
-        virtual void resume(error_code& ec = throws) = 0;
+        virtual void suspend(error_code& ec = throws) {};
+        virtual void resume(error_code& ec = throws) {};
```
Comment (hkaiser) [inline hpx/runtime/threads/detail/thread_pool_base.hpp]: Should the default implementations at least assert, if not throw an exception instead of silently ignoring the request?

File: tests/unit/resource/suspend_pool.cpp [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+int hpx_main(int argc, char* argv[])
+{
+    try
+    {
+        hpx::this_thread::get_pool()->suspend();
+        HPX_TEST_MSG(false, "Suspending is not possible on own pool");
+    }
+    catch (std::runtime_error const&)
+    {
+    }
```
Comment (hkaiser) [inline tests/unit/resource/suspend_pool.cpp]: Where is the `std::runtime_error` thrown? If this is thrown by HPX it should be changed to using an `hpx::exception` instead.

Either way, it looks like the exception is expected in this test case, if so, please add a bool checking that it was indeed thrown:

    bool exception_thrown = false;
    try { ... } catch(hpx::exception const& e) { exception_thrown = true; }
    HPX_TEST(exception_thrown);

File: tests/unit/resource/suspend_pool.cpp [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            {
+                fs.push_back(hpx::async(worker_exec, [](){}));
+            }
+
+            worker_pool.suspend();
+
+            // Can only suspend once all work is done
+            auto f = hpx::when_all(std::move(fs));
+            HPX_TEST(f.is_ready());
+
```
Comment (hkaiser) [inline tests/unit/resource/suspend_pool.cpp]: Is there a way for the `suspend()` API to return a `future<void>` allowing to synchronize with the fact that the pool actually has been suspended?

File: tests/unit/resource/suspend_pool.cpp [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                hpx::threads::policies::enable_elasticity);
+
+            std::unique_ptr<hpx::threads::detail::thread_pool_base> pool(
+                new hpx::threads::detail::scheduled_thread_pool<Scheduler>(
+                    std::move(scheduler), notifier, pool_index, pool_name, mode,
+                    thread_offset));
+
+            return pool;
+        });
+
```
Comment (hkaiser) [inline tests/unit/resource/suspend_pool.cpp]: I still think that the API for creating a new pool is too complicated to be exposed to mere mortal users. This needs to be streamlined somehow - but I understand that this is orthogonal to this PR and should be handled by a separate ticket.

File: hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp (hunk 1 above)
Comment (msimberg) [reply hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: `sleep(0)` (or `yield`) is a good idea. HPX thread I'm not sure about. It would require returning a future (if we want to know when it's done), and so far I've assumed this can be called from outside the runtime.

File: tests/unit/resource/suspend_pool.cpp (hunk 3 above)
Comment (msimberg) [reply tests/unit/resource/suspend_pool.cpp]: Hmm, yes, it is thrown by HPX. I did not realize it will be an `hpx::exception` then... Will update it the way you've suggested.

File: tests/unit/resource/suspend_pool.cpp (hunk 4 above)
Comment (msimberg) [reply tests/unit/resource/suspend_pool.cpp]: I suppose it could. However, I've again assumed that this can be called from outside the runtime in which case waiting for the future will not work. I don't have a strong case for calling it from outside the runtime but thought it's unnecessary to restrict it unless needed. This may be a reason to restrict it though.

File: tests/unit/resource/suspend_pool.cpp (hunk 5 above)
Comment (msimberg) [reply tests/unit/resource/suspend_pool.cpp]: 100% agreed on both points.

Comment (msimberg) [reply tests/unit/resource/suspend_pool.cpp]: Opened an issue: #3082

File: tests/unit/resource/suspend_pool.cpp (hunk 4 above)
Comment (hkaiser) [reply tests/unit/resource/suspend_pool.cpp]: Alternatively you could add a callback that is to be called once the operation is finished executing. That also would allow for splitting the API into two functions, one taking the callback (to be used when called from outside HPX), and one (internally) passing a `hpx::lcos::local::packaged_task` to the first API and returning a future extracted from the packaged_task (to be called from HPX threads). That should decouple the waiting from the actual context the suspend is invoked from.

File: hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp (hunk 1 above)
Comment (hkaiser) [reply hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: See here for how this could be handled: https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/util/detail/yield_k.hpp. In fact, you could use `yield_k()` instead of having your own loop.

File: hpx/runtime/threads/detail/thread_pool_base.hpp (hunk 2 above)
Comment (hkaiser) [reply hpx/runtime/threads/detail/thread_pool_base.hpp]: Thanks for this fix. Looks fine now.

File: tests/unit/resource/suspend_pool.cpp (hunk 3 above)
Comment (hkaiser) [reply tests/unit/resource/suspend_pool.cpp]: Thanks, LGTM.

File: tests/unit/resource/suspend_pool.cpp (hunk 4 above)
Comment (msimberg) [reply tests/unit/resource/suspend_pool.cpp]: Would you then spawn an OS thread or HPX thread depending on from where it's called? The downside is that there will be two functions, and the user has to know (finnd out) that only one of them can be called from outside the runtime (asserts help though, and returning a future is already a good sign it can't be called from the outside). I'll give it a shot in any case as it's a nice idea.

Comment (hkaiser) [reply tests/unit/resource/suspend_pool.cpp]: I think you can always spawn an HPX thread. Invoking the callback from an HPX thread should be a viable option, always.

Comment (msimberg) [reply tests/unit/resource/suspend_pool.cpp]: Ah, you're right. The problem is that the suspending thread will then run on the pool it's supposed to suspend. So either it gets stuck or it has to call the callback before it has fully suspended the pool. The latter feels wrong but could be good enough (it does first wait for all work to be done before really suspending). Maybe there's another way.

Comment (msimberg) [reply tests/unit/resource/suspend_pool.cpp]: Two corrections to my previous comments: I have to launch an OS thread in the case of suspending from non-HPX threads, and suspending from non-HPX threads is required for suspending the whole runtime...

I implemented a futurized and a callback version of `suspend` the way you suggested and the principle works well but I still need to tidy it up. The real question is: do you think it's useful that all of `suspend`/`resume`/`suspend_processing_unit`/`resume_processing_unit` have a futurized and a callback version? I think the answer is clearly yes if it's useful for `suspend`.

File: hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp (hunk 1 above)
Comment (msimberg) [reply hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: Updated to use `yield_k` - made most waits nice and short. Thanks for pointing that out.

File: hpx/runtime/threads/detail/io_service_thread_pool.hpp [hunk 6]
```diff
@@ -65,6 +65,13 @@ namespace hpx { namespace threads { namespace detail
         ///////////////////////////////////////////////////////////////////////
         void stop (std::unique_lock<compat::mutex>& l, bool blocking = true);
 
+        ///////////////////////////////////////////////////////////////////////
+        hpx::future<void> suspend(error_code& ec = throws);
```
Comment (hkaiser) [inline hpx/runtime/threads/detail/io_service_thread_pool.hpp]: The API returning a future does not need to take an `error_code` as all possible exceptions will be reported through the future mechanism. Do I miss something?

File: hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp [hunk 7]
```diff
@@ -316,13 +320,81 @@ namespace hpx { namespace threads { namespace detail
...
+        if (threads::get_self_ptr() && hpx::this_thread::get_pool() == this)
         {
+            HPX_THROWS_IF(ec, bad_parameter,
+                "scheduled_thread_pool<Scheduler>::suspend",
+                "cannot suspend a pool from itself");
             return;
         }
 
+        auto suspend_func_wrapper =
+            [this, callback{std::move(callback)}, ec]() mutable
```
Comment (hkaiser) [inline hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: This is C++14 only. We try to stay compatible with C++11. Please use the (new) `HPX_CAPTURE_FORWARD` macro instead of the direct initialization of the `callback` capture variable (see https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/config/lambda_capture.hpp).

File: hpx/runtime/threads/detail/io_service_thread_pool.hpp (hunk 6 above)
Comment (msimberg) [reply hpx/runtime/threads/detail/io_service_thread_pool.hpp]: I don't think you're missing anything. This part of what I meant when I said it needs tidying up (just wanted to have the basics working first to try it out).

File: hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp (hunk 7 above)
Comment (msimberg) [reply hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: Yep, I'm aware. Again needs tidying up. Sorry, should've been more explicit about what's still to do.

File: hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp [hunk 8]
```diff
@@ -1460,13 +1533,19 @@ namespace hpx { namespace threads { namespace detail
...
+
+            for (std::size_t k = 0;
+                 thread_num == hpx::get_worker_thread_num();
+                 ++k)
             {
-                hpx::this_thread::suspend();
+                util::detail::yield_k(k,
+                    "scheduled_thread_pool::remove_processing_unit_internal");
             }
         }
```
Comment (hkaiser) [inline hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: This loop looks like to be a good candidate for being extracted into a utility function (similar construct are used during shutdown as well, iirc).

File: hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp (hunk 7 above)
Comment (hkaiser) [reply hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: I C++11 we just copy into the capture :/ The only alternative is to use `bind`/`deferred_call` instead of the lambda.

File: hpx/runtime/threads/detail/io_service_thread_pool.hpp (hunk 6 above)
Comment (hkaiser) [reply hpx/runtime/threads/detail/io_service_thread_pool.hpp]: Fair enough. Things start looking very good now. Thanks for your work on this!

File: hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp (hunk 7 above)
Comment (msimberg) [reply hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: I had problems with the `packaged_task` as it's non-copyable (I need the move). But I think I can do without the `packaged_task` and just do an `hpx::async` for that case (IMO neither more complicated nor simpler). Otherwise it's like you say, `bind`/`deferred_call`.

File: tests/unit/resource/suspend_pool.cpp (hunk 4 above)
Comment (msimberg) [reply tests/unit/resource/suspend_pool.cpp]: @hkaiser Besides technical details, do you have an opinion on the last question? (In case this slipped past you...)

File: hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp (hunk 8 above)
Comment (msimberg) [reply hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: You're right. The loops at shutdown are at least similar, I'll see what common functionality I can get out of there.

Comment (msimberg) [reply hpx/runtime/threads/detail/scheduled_thread_pool_impl.hpp]: I've pulled out a `yield_while` function into `yield_k.hpp`, which takes a predicate for terminating the loop and some optional arguments (same as in `yield_k`). It doesn't really make the code shorter but I think the intention is clearer, and `k` is now hidden which I think is good.
